From 2cf63f7925eaf00f184e532ca6fd30a224a7d668 Mon Sep 17 00:00:00 2001
From: Ken Tseng <ken.tseng@realtek.com>
Date: Tue, 20 Jan 2026 15:25:51 +0800
Subject: [PATCH] Bluetooth: Support Realtek RTL8822ES UART BT

- rtk_rfkill: add voltage detection by LSADC
-- export struct gpio_desc *bt_reset
---
 drivers/bluetooth/rtk_rfkill.c                | 79 +++++++++++--------

diff --git a/drivers/bluetooth/rtk_rfkill.c b/drivers/bluetooth/rtk_rfkill.c
index a044ffa63cd54..f3f09f2fbf028 100644
--- a/drivers/bluetooth/rtk_rfkill.c
+++ b/drivers/bluetooth/rtk_rfkill.c
@@ -30,6 +30,8 @@
 
 #define LSADC0_PAD0 0x900
 #define ISO_PFUNC21 0x74
+#define ADC_VAL0_MASK	GENMASK(7, 0)
+#define ADC_VAL0	0x94
 
 #define ISO_DBG_STATUS 0x190
 #define ISO_PFUNC0 0x2c
@@ -46,12 +48,15 @@ enum voltage_config {
 
 static struct rfkill *bt_rfk;
 static const char bt_name[] = "bluetooth";
-static struct gpio_desc *bt_reset;
+struct gpio_desc *bt_reset = NULL;
+EXPORT_SYMBOL_GPL(bt_reset);
 
-struct pinctrl *pinctrl;
-struct pinctrl_state *pins_default;
-struct pinctrl_state *pins_1v8;
-struct pinctrl_state *pins_3v3;
+static struct regmap *iso_dbg = NULL;
+static struct regmap *pinctrl_base = NULL;
+static struct pinctrl *pinctrl = NULL;
+static struct pinctrl_state *pins_default = NULL;
+static struct pinctrl_state *pins_1v8 = NULL;
+static struct pinctrl_state *pins_3v3 = NULL;
 
 static int bluetooth_set_power(void *data, bool blocked)
 {
@@ -92,13 +97,43 @@ static void rfkill_gpio_deinit(void)
 	gpiod_put(bt_reset);
 }
 
+void rtk_voltage_detect_and_set(int method, int mode)
+{
+	if (mode == DETECT) {
+	if (method == 0 && pinctrl_base) {
+		unsigned int val = 0x0;
+		val = BIT(6);
+		regmap_update_bits_base(pinctrl_base, ISO_DBG_STATUS, val, val,
+					NULL, false, true);
+		mdelay(100);
+		regmap_read(pinctrl_base, ISO_DBG_STATUS, &val);
+
+		mode = (val & BIT(3)) ? FIX_3V3 : FIX_1V8;
+	}
+	else if (method == 1 && iso_dbg) {
+		u32 reg, val_mask;
+		regmap_read(iso_dbg, LSADC0_PAD0, &reg);
+		val_mask = reg & ADC_VAL0_MASK;
+
+		mode = (val_mask >= ADC_VAL0) ? FIX_3V3 : FIX_1V8;
+	}
+	}
+	if (mode == FIX_1V8) {
+		if (pins_1v8)
+			pinctrl_select_state(pinctrl, pins_1v8);
+	}
+	else if (mode == FIX_3V3) {
+		if (pins_3v3)
+			pinctrl_select_state(pinctrl, pins_3v3);
+	}
+}
+EXPORT_SYMBOL_GPL(rtk_voltage_detect_and_set);
+
 static int rfkill_bluetooth_probe(struct platform_device *pdev)
 {
 	struct device_node *node = pdev->dev.of_node;
 	struct device *dev = &pdev->dev;
 	bool default_state = true;
-	struct regmap *iso_dbg;
-	unsigned int val = 0x0;
 	const u32 *prop;
         int size;
 	int rc = 0;
@@ -117,6 +152,9 @@ static int rfkill_bluetooth_probe(struct platform_device *pdev)
 		dev_dbg(dev, "no pinctrl\n");
 		mode = DETECT_NONE;
 	} else {
+		iso_dbg = syscon_regmap_lookup_by_phandle(node, "realtek,iso");
+		pinctrl_base = syscon_regmap_lookup_by_phandle(
+					node, "realtek,pinctrl");
 	        pins_default = pinctrl_lookup_state(pinctrl, PINCTRL_STATE_DEFAULT);
 		if (IS_ERR(pins_default)) {
 			dev_warn(dev, "could not get default state\n");
@@ -140,32 +178,7 @@ static int rfkill_bluetooth_probe(struct platform_device *pdev)
         if (prop)
                 mode = of_read_number(prop, 1);
 
-	if (mode == DETECT) {
-		iso_dbg = syscon_regmap_lookup_by_phandle(
-					node, "realtek,pinctrl");
-
-		val = BIT(6);
-		regmap_update_bits_base(iso_dbg, ISO_DBG_STATUS, val, val,
-					NULL, false, true);
-		mdelay(100);
-		regmap_read(iso_dbg, ISO_DBG_STATUS, &val);
-
-		if(val & BIT(3)) {	/* 3.3v */
-			if (pins_3v3)
-				pinctrl_select_state(pinctrl, pins_3v3);
-		} else {			/* 1.8v */
-			if (pins_1v8)
-				pinctrl_select_state(pinctrl, pins_1v8);
-		}
-	}
-	else if (mode == FIX_1V8) {
-		if (pins_1v8)
-			pinctrl_select_state(pinctrl, pins_1v8);
-	}
-	else if (mode == FIX_3V3) {
-		if (pins_3v3)
-			pinctrl_select_state(pinctrl, pins_3v3);
-	}
+	rtk_voltage_detect_and_set(0, mode);
 
 	bt_rfk = rfkill_alloc(bt_name, &pdev->dev, RFKILL_TYPE_BLUETOOTH,
 				&rfkill_bluetooth_ops, NULL);
-- 
2.34.1

